[14.0][ADD] invader_payment_adyen_klarna

acsonefho opened the pull request:
Add some information related to Klarna on the request to do the payment.

Depends on https://github.com/shopinvader/odoo-shopinvader-payment/pull/57

**But also some refactor:**
- invader_payment: invader.payable: the format amount depends on the transaction (parameter) and should be available on the payable (not only on the service);
- invader_payment adyen + stripe: define a decimal precision used to rounding during format_amount (instead of hard-code 2 digits);
- invader_payment_adyen_klarna: some value are optionnal in odoo so instead of sending False, sent empty string (then if the external Adyen lib is not ok with data, that's another stuff. But format is ok);
- service in invader_payment_adyen_klarna: refactor to use invader.payable abstraction;
- create abstract module invader_payment_adyen_klarna_ for invoice and sale + implementation;
- shopinvader_invoice_payment: move some abstraction into invader_invoice_payment.

I did the refactor into a dedicated commit. I also did this into this PR because it's quite hard to do some test with multi-pr (with a lot of abstraction) and then this PR will depends on many other PR.

acsonefho committed:
[ADD] invader_payment_adyen_klarna

rousseldenis reviewed (approved):
Needs to fix #60

Cedric-Pigeon commented on invader_payment_adyen_klarna/services/payment_adyen.py:
```diff
@@ -0,0 +1,131 @@
+        return {
+            "id": line.id,
+            "quantity": line.product_uom_qty,
+            "taxPercentage": line.tax_id.amount,
```
you should call self._get_formatted_amount

Cedric-Pigeon commented on invader_payment_adyen_klarna/services/payment_adyen.py:
```diff
@@ -0,0 +1,131 @@
+            "quantity": line.product_uom_qty,
+            "taxPercentage": line.tax_id.amount,
+            "description": line.display_name,
+            "amountIncludingTax": line.price_total,
```
same here

Cedric-Pigeon commented on invader_payment_adyen_klarna/services/payment_adyen.py:
```diff
@@ -0,0 +1,131 @@
+        return {
+            "id": line.id,
+            "quantity": line.quantity,
+            "taxPercentage": first(line.tax_ids).amount,
```
same here

Cedric-Pigeon commented on invader_payment_adyen_klarna/services/payment_adyen.py:
```diff
@@ -0,0 +1,131 @@
+            "quantity": line.quantity,
+            "taxPercentage": first(line.tax_ids).amount,
+            "description": line.display_name,
+            "amountIncludingTax": line.price_total,
```
same here

rousseldenis commented:
@acsonefho @Cedric-Pigeon It could be great to realign PR's in order to make them merged.

acsonefho committed:
[FIX] invader_payment_adyen: missing deps about base_rest (used into abstract service)

acsonefho commented:
@shopinvader/shopinvader-maintainers Can I have review/merge please?
Already in production so functionally working

PierrickBrun reviewed (commented):
There is a lot of duplicated code for invoices and sale orders that could be refactored.

I think using the abstract "payable" object would be a good solution to that and be more aligned with the rest of the repo.

lmignon commented on invader_payment_adyen_klarna/services/payment_adyen.py:
```diff
@@ -0,0 +1,145 @@
+
+    def _klarna_partner_info(self, partner):
+        return {
+            "city": partner.city,
+            "country": partner.country_id.code,
+            "houseNumberOrName": "",
+            "postalCode": partner.zip,
+            "street": partner.street,
```
@acsonefho Be careful since none of these fields are mandatory into odoo. Therefore, you must avoid to return False in place of an empty string...  `"city": partner.city or ""`

lmignon reviewed (changes requested):
A little change....

acsonefho commented:
> There is a lot of duplicated code for invoices and sale orders that could be refactored.
> 
> I think using the abstract "payable" object would be a good solution to that and be more aligned with the rest of the repo.

You're totally right!
It was a discussion that I had today. I'll do it soon

acsonefho committed:
[IMP] Refactor:
- invader_payment: invader.payable: the format amount depends on the transaction (parameter) and should be available on the payable (not only on the service);
- invader_payment adyen + stripe: define a decimal precision used to rounding during format_amount (instead of hard-code 2 digits);
- invader_payment_adyen_klarna: some value are optionnal in odoo so instead of sending False, sent empty string (then if the external Adyen lib is not ok with data, that's another stuff. But format is ok);
- service in invader_payment_adyen_klarna: refactor to use invader.payable abstraction;
- create abstract module invader_payment_adyen_klarna_ for invoice and sale + implementation;
- shopinvader_invoice_payment: move some abstraction into invader_invoice_payment.

acsonefho commented:
@shopinvader/shopinvader-maintainers Some review here please? Huge one...

Already in production and working properly.

Cedric-Pigeon commented:
/ocabot merge patch
